### ISSUE TITLE
Gradient accumulation done wrong

### ISSUE THREAD

**Issue by shkarupa-alex** [keras-team/keras#19990]
In my opinion gradient accumulation should imitate large batch size (if model changes it's weight only through optimizer - not like in BatchNorm).
So here are 3 tests all with batch_size=12+gradient_accumulation_steps=1 vs batch_size=4+gradient_accumulation_steps=3 https://colab.research.google.com/drive/11MdIRQVQoSobYuev-1bCe7EuoEpU0ghU?usp=sharing 

SGD without weight_decay - works as expected.
SGD with weight_decay - fails, but ok if weight_decay divided by gradient_accumulation_steps
Adam - always fails.

In my opinion issue with SGD occuires because weight decay is applied in every step, but should only after real variable update. This is a performance issue too.

Adam issue as i think has one more reason: this optimizers uses self.iterations which is different in every accumulation step (and shoul use self.iterations // gradient_accumulation_steps).
